File: algo/Basic Algorithm/STL/Priority Queue/大根堆.py
def up(x):
    if(x==1):
        return
    if(heap[x]>heap[x//2]):
        heap[x],heap[x//2]=heap[x//2],heap[x]
        up(x//2)

def down(x):
    global tot
    if(x*2>tot):
        return
    son=x*2
    if(x*2+1<=tot and heap[x*2+1]>heap[x*2]):
        son+=1
    if(heap[son]>heap[x]):
        heap[x],heap[son]=heap[son],heap[x]
        down(son)

def add(x):
    global tot
    heap.append(x)
    tot+=1
    up(tot)

def pop():
    global tot
    heap[1]=heap[tot]
    heap.pop(tot)
    tot-=1
    down(1)

n=int(input())
tot=0
heap=[0]
for i in range(n):
    opt=input().split()
    if(opt[0]=='push'):
        x=int(opt[1])
        add(x)
    else:
        if(tot==0):
            print('empty')
        else:
            print(heap[1])
            pop()

File: algo/Data Stucture/Heap/大根堆.cpp
/*Author：AustinJiang
题目：大根堆
时间复杂度：O(n*log(n))
算法：堆，优先队列 */
#include<bits/stdc++.h>
#define ll long long
#define pb push_back
#define mp make_pair
#define fir first
#define sec second
#define endl "\n"
#define random(a,b) rand()%(b-a+1)+a
#define PI pair<int,int>
#define VI vector<int>
#define VPI vector<PI>
#define PQ priority_queue
using namespace std;
const int INF=0x3f3f3f3f;
const int N=1e5+10;

int n,tot,heap[N];

inline void push_up(int x){
	if(x==1) return;
	if(heap[x]>heap[x/2]){
		swap(heap[x],heap[x/2]);
		push_up(x/2);
	}
}

inline void push_down(int x){
	int son=x*2;
	if(son>tot) return;
	if(son+1<=tot&&heap[son+1]>heap[son]) son++;
	if(heap[x]<heap[son]){
		swap(heap[x],heap[son]);
		push_down(son);
	}
}

inline void push(int x){
	heap[++tot]=x;
	push_up(tot);
}

inline void pop(){
	swap(heap[1],heap[tot--]);
	push_down(1);
}

signed main(){
	ios::sync_with_stdio(false);
	cin.tie(0);
	cin>>n;
	for(int i=1;i<=n;i++){
		string opt;
		int x;
		cin>>opt;
		if(opt=="push") cin>>x,push(x);
		else if(!tot) cout<<"empty"<<endl;
		else cout<<heap[1]<<endl,pop();
	}
	return 0;
}

